src/ripple/app/data/tests/SociDB.test.cpp

```diff
@@ -295,11 +295,77 @@ class SociDB_test final : public TestSuite
                 remove (dbPath);
         }
     }
+    void testSQLiteDeleteWithSubselect()
+    {
+        testcase ("deleteWithSubselect");
+        BasicConfig c;
+        setupSQLiteConfig (c, getDatabasePath ());
+        SociConfig sc (c, "SociTestDB");
+        soci::session s;
+        sc.open (s);
+        const char* dbInit[] = {
+            "BEGIN TRANSACTION;",
+            "CREATE TABLE Ledgers (                         \
+                LedgerHash      CHARACTER(64) PRIMARY KEY,  \
+                LedgerSeq       BIGINT UNSIGNED             \
+            );",
+            "CREATE INDEX SeqLedger ON Ledgers(LedgerSeq);",
+
+            "CREATE TABLE Validations   (                   \
+                LedgerHash  CHARACTER(64)                   \
+            );",
+            "CREATE INDEX ValidationsByHash ON              \
+                Validations(LedgerHash);",
+            "END TRANSACTION;"};
+        int dbInitCount = std::extent<decltype(dbInit)>::value;
+        for (int i = 0; i < dbInitCount; ++i)
+        {
+            s << dbInit[i];
+        }
+        char lh[65];
+        memset (lh, 'a', 64);
+        lh[64] = '\0';
+        int toIncIndex = 63;
+        int const numRows = 16;
+        for (int i = 0; i < numRows; ++i)
+        {
+            ++lh[toIncIndex];
+            if (lh[toIncIndex] == 'z')
+                --toIncIndex;
+            std::string slh(lh);
+            s << "INSERT INTO Ledgers (LedgerHash, LedgerSeq) VALUES "
+                 "(:lh, :li);", soci::use (slh), soci::use (i);
+            s << "INSERT INTO Validations (LedgerHash) VALUES "
+                 "(:lh);", soci::use (slh);
+        }
+        std::vector<int> ledgersLS (numRows * 2);
+        std::vector<std::string> validationsLH (numRows * 2);
+        s << "SELECT LedgerSeq FROM Ledgers;", soci::into (ledgersLS);
+        s << "SELECT LedgerHash FROM Validations;", soci::into (validationsLH);
+        expect (ledgersLS.size () == numRows &&
+                validationsLH.size () == numRows);
+        s << "DELETE FROM Validations WHERE LedgerHash IN "
+             "(SELECT Ledgers.LedgerHash FROM Validations JOIN Ledgers ON "
+             "Validations.LedgerHash=Ledgers.LedgerHash WHERE Ledgers.LedgerSeq < :num);",
+            soci::use (numRows / 2);
+        validationsLH.resize (numRows * 2);
+        s << "SELECT LedgerHash FROM Validations;", soci::into (validationsLH);
+        expect (validationsLH.size () == numRows / 2);
```

Comment:
Checking the size of `validationsLH` here is great. One of two things could make it better:

1. Check the values. You generated them earlier, so you could (potentially) keep a local copy and compare them.
2. Query `Validations JOIN Ledger` and verify that the expected ones were deleted.
```
s << "SELECT Ledgers.LedgerSeq FROM Validations JOIN Ledgers ON "
"Validations.LedgerHash=Ledgers.LedgerHash WHERE Ledgers.LedgerSeq < :num",
soci::use(numRows / 2), soci::into(ledgersLS);
expect(ledgersLS.size() == 0);
```
Either way, the idea is to verify that the _right_ rows were deleted, not just the right _number_ or rows.